### ISSUE TITLE
H5P plugin update: language string issues (version 1.20.0 (2019101700))

### ISSUE THREAD

**Issue by gemguardian** [h5p/moodle-mod_hvp#313]
I am not sure if it is just an issue with the DUTCH language strings or that this in all other local changes except English. 

**Though** I am sure that this issue is **only** caused when you **update** the plugin **not** when you do a **fresh** installation. I had **no** issues with a fresh installation and the dutch language strings

I have updated my H5P plugin to the latest version:  1.20.0 (2019101700)

After updating my plugin in Moodle 3.5 I got the 
![string1_error](https://user-images.githubusercontent.com/16559999/68324636-0ecd0d80-00c8-11ea-975f-859cf6683af0.png)

I did a fresh H5P installation on a local 3.7, setted the language NL and I had **no** error. 

Checking **both** databases I noticed that in the new installation the NL stringid had **no** capital, but in the **updated** installation it did. 
![db_difference_hvp](https://user-images.githubusercontent.com/16559999/68324944-bc402100-00c8-11ea-9227-0a0aaa9d2823.png)

I did a cache clearning - no changes. 
I changed the string in the database > the error jump over to a different string. 

In total I identified the following strings as an issue:

- reusedescription
- contentcopied
- reusecontent
after changing them in the updated database the errors where gone. 

I figured that it would be best to report this here, in case others get this this error. I also mentioned it in AMOS to the dutch content while giving the new translated strings, I am not sure if he can do something about it. 

Happy debugging 

Gemma

**Comment by gemguardian** [comment]
@thomasmars I just noticed you posted a bug fix https://github.com/h5p/h5p-moodle-plugin/pull/308 16 days ago. Would that fix my isseu? (I am not a coder that's why I wonder)

**Comment by thomasmars** [comment]
Hi Gemma,
Thanks for the detailed report.
The bug fix should not be related to what you're seeing.

I believe this is because we changed capitalization of the strings in https://github.com/h5p/h5p-moodle-plugin/commit/e3e340043acbbda0521718738c9b04e35a3ead8a. Since the AMOS system does not seem to be case insensitive this is causing conflicts if you had already registered the languages in your database and then updated to the new version.
I suggest trying to clear the language cache (https://docs.moodle.org/37/en/Developer_tools#Purge_all_caches) to see if this resolves the problem.
I'll be looking into alternative ways to resolve this.

**Comment by gemguardian** [comment]
Hi thomas

Clearing the language cache did not resolve the issue. One of the first things i tried:)



Verzonden vanaf mijn Samsung Galaxy-smartphone.


-------- Oorspronkelijk bericht --------
Van: Thomas <notifications@github.com>
Datum: 07-11-19 09:50 (GMT+01:00)
Aan: h5p/h5p-moodle-plugin <h5p-moodle-plugin@noreply.github.com>
Cc: Gemma Lesterhuis <gemma@ltnc.nl>, Author <author@noreply.github.com>
Onderwerp: Re: [h5p/h5p-moodle-plugin] H5P plugin update: language string issues (#313)


Hi Gemma,
Thanks for the detailed report.
The bug fix should not be related to what you're seeing.

I believe this is because we changed capitalization of the strings in e3e3400<https://github.com/h5p/h5p-moodle-plugin/commit/e3e340043acbbda0521718738c9b04e35a3ead8a>. Since the AMOS system does not seem to be case insensitive this is causing conflicts if you had already registered the languages in your database and then updated to the new version.
I suggest trying to clear the language cache (https://docs.moodle.org/37/en/Developer_tools#Purge_all_caches) to see if this resolves the problem.
I'll be looking into alternative ways to resolve this.

**Comment by Urpokarhu1** [comment]
Hello,

I have ran in to same problem with mod_hvp 1.20 I cannot anymore edit/open any language packs

**Comment by thomasmars** [comment]
Thanks.
Am I right in assuming that you both have checked out the language locally (at your site) through admin/tool/customlang/index.php?lng=nl ? (or other language).

Since the translations strings have changed, they are no longer compatible, so I think you have to delete or amend the incompatible translation strings.

I wasn't able to reproduce this in Moodle 3.5.8+ (Build: 20191016). Could you let me know which version of Moodle you're using and which database? I'll try again with your version and create some steps you can follow to resolve the issue if I'm able to reproduce and fix it.

**Comment by Urpokarhu1** [comment]
Hello,

Yes i'm trying to edit f.ex english language from 

admin/tool/customlang/index.php?lng=en

when I get the error message, we are using MariaDB

**Comment by Urpokarhu1** [comment]
Hello,

You might need to edit H5P strings before updating to mod_hvp 1.20 to reproduce this error

**Comment by thomasmars** [comment]
I was able to reproduce this with MariaDB, I guess it requires case insensitive uniqueness in indexes, which postgres does not.

To fix the issue you can choose between the following approaches:
- If you don't have any custom translations you need to keep you can truncate (delete all rows) the `mdl_tool_customlang` table, then you'll fetch all the latest translations directly from AMOS, instead of trying to merge them with what you have locally
- If you have custom translations you want to keep you can:
  - Delete the keys in your database that matches the keys in this commit: https://github.com/h5p/h5p-moodle-plugin/commit/e3e340043acbbda0521718738c9b04e35a3ead8a
  - Or if you've made custom translations for the key values you can change the keys in your database that matches the keys in the commit to be lowercase

I'm sorry for the inconvenience, I'm also contacting Moodle to hear if this can be amended through the AMOS system.

**Comment by koenr** [comment]
I can confirm that there have been more problems using capitals in string names in Amos, although I can't remember me the exact case. Particularly when using a disk file system on the server that is not case sensitive (Windows servers (or desktops as a test machine), Apple).  Reason there was that the custom lang tool in Moodle creates files. NTFS is not case sensitive, causing one file for two different strings.

**Comment by gemguardian** [comment]
@thomasmars sorry for not responding yesterday but was on the road. 

Thank you for looking in to this issue. 

It is not only MariaDB that gives this issue, but also mysql. 
And only when you update the H5P plugin, not if you do a do a first installation of H5P. 

I can confirm that "Or if you've made custom translations for the key values you can change the keys in your database that matches the keys in the commit to be lowercase" solution works as a charm.

**Comment by raad** [comment]
Here's what I used as a workaround on a Moodle 3.5.8 with MySQL instance:

```sql
update prefix_tool_customlang
set    stringid  = lower(stringid)
where  componentid = (select id from prefix_tool_customlang_components where name = 'mod_hvp')
and    lower(stringid) not like binary(stringid)
and    stringid not like 'license%'
```
Don't forget to change "prefix" to your Moodle database prefix if you have one.

**Comment by gemguardian** [comment]
@raad thank you for the solution. 

To add to this tracker, the update that is been released yesterday on Moodle.org (1.20.1 (2019111300)) has similar issues with the new language strings.

**Comment by gemguardian** [comment]
small correction >> if you **have** installed version 1.20.0 (2019101700) and **have not** changed the language strings yet, and **then** update to the new version (1.20.1 (2019111300)) there will be more language strings to "fix". 

There are **no** issues when:

-  you update directly to (1.20.1 (2019111300)) and skipping version  (1.20.1 (2019111300)
ór 

- when you **have** fixed the language strings and then update to (1.20.1 (2019111300))

**Comment by jgg9026** [comment]
`        $sql = "UPDATE {tool_customlang}
                   SET stringid = LOWER(stringid)
                 WHERE componentid = (SELECT id FROM {tool_customlang_components} customlang WHERE customlang.name = 'mod_hvp')
                       AND " . $DB->sql_like("LOWER(stringid)", "stringid", true, true, true) .
                     " AND stringid NOT LIKE 'license%'";
        $DB->execute($sql);`

What would happen if we run that in a PostgreSQL environment?

**Comment by dezglidden** [comment]
Experienced this also, on MySQL 5.7.24 

My initial thought was this is a result of the change in commit e3e340043acbbda0521718738c9b04e35a3ead8a and that adding an upgrade step in upgrade.php would be the best approach, given that everyone that's followed the upgrade step identified in https://github.com/h5p/h5p-moodle-plugin/issues/313#issuecomment-553780931 will have the same issue

**Comment by thomasmars** [comment]
Yeah, should be fine adding an upgrade.php step. We would love to receive a pull request for this.

**Comment by dezglidden** [comment]
Cool, I shall make time for a patch this week

**Comment by dezglidden** [comment]
So I attempted to replicate this on a fresh install using 1.18 and upgrading to 1.20. I didn't encounter the dmlwriteexception issue, instead I found that both lowercase and camel case strings co-existed as separate database entries (reusecontent and reuseContent). Seems a universal patch isn't straight-forward, since potentially converting to lowercase could leave duplicate entries for several strings. I also found a number of extra camel case strings during the new install that were not present in the codebase we discovered this issue on.

**Comment by emmarichardson** [comment]
FYI - this is also happening with en_us language string...

**Comment by cidor361** [comment]
I had this problem (mod_hvp 1.20) with russian languages after update moodle (from 3.5.6 to 3.9.1+). 
![hvp_error](https://user-images.githubusercontent.com/36622579/92447555-0c43b480-f1c0-11ea-84fd-4bae2f6c601a.jpg)
I applied changes in commit e3e3400 (didn't help). Then i noticed that in db (10.1.26 MariaDB) the same words contain upper case letters.
![reuseContect_mooc](https://user-images.githubusercontent.com/36622579/92447657-2bdadd00-f1c0-11ea-9288-24339f7ba5dd.png)
Below are the queries that fix the upper case in DB (after that the problem did not appear anymore):
`UPDATE mdl_tool_customlang SET stringid='reusecontent' WHERE stringid='reuseContent';
UPDATE mdl_tool_customlang SET stringid='reusedescription' WHERE stringid='reuseDescription';
UPDATE mdl_tool_customlang SET stringid='contentcopied' WHERE stringid='contentCopied';
UPDATE mdl_tool_customlang SET stringid='connectionlost' WHERE stringid='connectionLost';
UPDATE mdl_tool_customlang SET stringid='connectionreestablished' WHERE stringid='connectionReestablished';
UPDATE mdl_tool_customlang SET stringid='resubmitscores' WHERE stringid='resubmitScores';
UPDATE mdl_tool_customlang SET stringid='offlinedialogheader' WHERE stringid='offlineDialogHeader';
UPDATE mdl_tool_customlang SET stringid='offlinedialogbody' WHERE stringid='offlineDialogBody';
UPDATE mdl_tool_customlang SET stringid='offlinedialogretrymessage' WHERE stringid='offlineDialogRetryMessage';
UPDATE mdl_tool_customlang SET stringid='offlinedialogretrybuttonlabel' WHERE stringid='offlineDialogRetryButtonLabel';
UPDATE mdl_tool_customlang SET stringid='offlinesuccessfulsubmit' WHERE stringid='offlineSuccessfulSubmit';
UPDATE mdl_tool_customlang SET stringid='fileexceedsmaxsize' WHERE stringid='fileExceedsMaxSize';
UPDATE mdl_tool_customlang SET stringid='unpackedfilesexceedsmaxsize' WHERE stringid='unpackedFilesExceedsMaxSize';
UPDATE mdl_tool_customlang SET stringid='couldnotreadfilefromzip' WHERE stringid='couldNotReadFileFromZip';
UPDATE mdl_tool_customlang SET stringid='couldnotparsejsonfromzip' WHERE stringid='couldNotParseJSONFromZip';
UPDATE mdl_tool_customlang SET stringid='couldnotparsepostdata' WHERE stringid='couldNotParsePostData';`

**Comment by bfriesenvcc** [comment]
Sorry all. I'm confused about what the final working solution is for this. 

Based on this advice:

"If you have custom translations you want to keep you can:
- Delete the keys in your database that matches the keys in this commit: e3e3400
- Or if you've made custom translations for the key values you can change the keys in your database that matches the keys in the commit to be lowercase"

I had our I.T. guy make back-end changes to the tool_customlang table so that there were no longer duplicates (Sample shown here. id 549 = mod_hvp and id 610 = core_h5p), yet I still am unable to open up the language pack. 

Should I get I.T. to delete all of the id=549 mod_hvp entries?

We are using H5P mod-hvp 1.21.0 build 2020080400
<img width="238" alt="LanguagePack" src="https://user-images.githubusercontent.com/8193384/105406055-c45a8280-5be0-11eb-8820-067bec36392d.PNG">

If you can help me resolve this, I will send you a virtual hug.

**Comment by thomasmars** [comment]
What error(s) are you getting ?

**Comment by bfriesenvcc** [comment]
This is the latest debugging report. However, when I look into the table, there is no 'licenceCC40' entry, only a 'licencecc40' (Note: I had my I.T. guy backend change all uppercase to lowercase. 
<img width="515" alt="Errors" src="https://user-images.githubusercontent.com/8193384/105735210-48b24b80-5ee8-11eb-9cca-b8f11dd37707.PNG">

**Comment by UncleRaymondo** [comment]
Issue still exists in Moodle 4.1.1, with HVP 1.23 and MySQL 8.0.32 - unable to open any language customisations now even though I was able to before, given that the site is in 4 languages. Language customisations is critical to our operation; how do I resovle this urgently?

**Comment by dezglidden** [comment]
In the end for us, having only experienced this for a single client, IIRC I headed straight to the database and then ran SQL statements as mentioned by raad further up the thread. It was by far the quickest and simplest solution across all their Moodle sites, testing locally and on staging first of course 😅

**Comment by UncleRaymondo** [comment]
@dezglidden yeah I tried those yesterday... spent 4 hours cleaning the mess as it literally killed the LMS... I'm not sure if the commands will work on Moodle 4.1.1; I kind of need to try them again, but it never worked yesterday... :(

**Comment by bfriesenvcc** [comment]
> Issue still exists in Moodle 4.1.1, with HVP 1.23 and MySQL 8.0.32 - unable to open any language customisations now even though I was able to before, given that the site is in 4 languages. Language customisations is critical to our operation; how do I resovle this urgently?

Here's how I get around it while it's still broken: https://yourmoodlesite/admin/tool/customlang/edit.php?lng=en

**Comment by UncleRaymondo** [comment]
> > Issue still exists in Moodle 4.1.1, with HVP 1.23 and MySQL 8.0.32 - unable to open any language customisations now even though I was able to before, given that the site is in 4 languages. Language customisations is critical to our operation; how do I resovle this urgently?
> 
> Here's how I get around it while it's still broken: https://yourmoodlesite/admin/tool/customlang/edit.php?lng=en

Cool... I'll give it a go later today :)

**Comment by lucaboesch** [comment]
@gemguardian were you able to track the issue down?
Could you please check if this ticket could be closed from your point of view?

**Comment by otacke** [comment]
@gemguardian ping

**Comment by otacke** [comment]
@gemguardian pong

**Comment by otacke** [comment]
@gemguardian ping pong.

**Comment by otacke** [comment]
@lucaboesch Same here: Is there a better way to reach out to @gemguardian?

**Comment by gemguardian** [comment]
@otacke I have not seen this issue anymore - but I have to be honest that most of my clients moved away from the addon to the core version.

**Comment by otacke** [comment]
@gemguardian Thanks here as well. I don't know if @lucaboesch wants to investigate this further (Luca, could you reopen the issue in case you do and find anything?), but I'll close it then for now. Sounded like some migration issue, and nobody seems to have reported something similar for 5.x instances.
